File: packages/scanner-py/src/aegis/checks/reputation.py
"""Reputation checks backed by the GoPlus token-security data source.

These consume a pre-fetched ``GoPlusReport`` stashed in ``ctx.cache["goplus"]``
by the engine. They are strictly additive: when GoPlus has no data (new token,
unsupported chain, or the source was disabled) every check returns nothing and
the on-chain battery stands alone. GoPlus flags never set the verdict directly —
they contribute score and evidence like any other finding.
"""

from __future__ import annotations

from ..models import CheckResult, Severity
from ..sources.goplus import GoPlusReport
from .base import Context


def _report(ctx: Context) -> GoPlusReport | None:
    data = ctx.cache.get("goplus")
    return data if isinstance(data, GoPlusReport) else None


class ReputationHoneypotCheck:
    """GoPlus honeypot + trading-tax signal, independent of our eth_call sim."""

    id = "REP-HONEYPOT"

    async def run(self, ctx: Context) -> list[CheckResult]:
        rep = _report(ctx)
        if rep is None:
            return []
        out: list[CheckResult] = []

        if rep.is_honeypot:
            out.append(
                CheckResult(
                    check=self.id,
                    severity=Severity.DANGER,
                    score=90,
                    title="GoPlus flags this as a honeypot",
                    detail=(
                        "GoPlus Security reports is_honeypot=1 for this token. Independent of "
                        "our own transfer simulation, external scanners see sells being blocked "
                        "or reverted. Do not buy."
                    ),
                    evidence={"source": "goplus", "is_honeypot": "1"},
                )
            )

        buy = rep.buy_tax if rep.buy_tax is not None else 0.0
        sell = rep.sell_tax if rep.sell_tax is not None else 0.0
        worst = max(buy, sell)
        if worst >= 0.50:
            out.append(
                CheckResult(
                    check=self.id,
                    severity=Severity.DANGER,
                    score=80,
                    title=f"Extreme trading tax ({worst:.0%})",
                    detail=(
                        f"GoPlus reports buy tax {buy:.0%} / sell tax {sell:.0%}. A tax this high "
                        "means most of your position is confiscated on entry or exit — a common "
                        "soft-rug pattern."
                    ),
                    evidence={"source": "goplus", "buy_tax": str(buy), "sell_tax": str(sell)},
                )
            )
        elif worst >= 0.10:
            out.append(
                CheckResult(
                    check=self.id,
                    severity=Severity.WARN,
                    score=25,
                    title=f"High trading tax ({worst:.0%})",
                    detail=(
                        f"GoPlus reports buy tax {buy:.0%} / sell tax {sell:.0%}. Factor this into "
                        "your expected slippage — it is charged on top of AMM price impact."
                    ),
                    evidence={"source": "goplus", "buy_tax": str(buy), "sell_tax": str(sell)},
                )
            )
        elif worst > 0.0:
            out.append(
                CheckResult(
                    check=self.id,
                    severity=Severity.INFO,
                    score=0,
                    title=f"Small trading tax ({worst:.1%})",
                    detail=f"GoPlus reports buy tax {buy:.1%} / sell tax {sell:.1%}.",
                    evidence={"source": "goplus"},
                )
            )
        return out


class ReputationPermissionsCheck:
    """Dangerous admin capabilities GoPlus surfaces from bytecode analysis."""

    id = "REP-PERMISSIONS"

    async def run(self, ctx: Context) -> list[CheckResult]:
        rep = _report(ctx)
        if rep is None:
            return []
        out: list[CheckResult] = []

        flags = [
            (
                rep.can_take_back_ownership,
                Severity.DANGER,
                60,
                "Ownership can be reclaimed",
                (
                    "GoPlus reports can_take_back_ownership=1: a renounced owner can be restored, "
                    "so any 'ownership renounced' claim is not final."
                ),
            ),
            (
                rep.owner_change_balance,
                Severity.DANGER,
                70,
                "Owner can modify balances",
                (
                    "GoPlus reports owner_change_balance=1: the owner can arbitrarily change holder "
                    "balances, effectively minting or zeroing your position."
                ),
            ),
            (
                rep.hidden_owner,
                Severity.DANGER,
                55,
                "Hidden owner detected",
                (
                    "GoPlus reports hidden_owner=1: privileged control is retained through a "
                    "non-obvious mechanism even if owner() looks renounced."
                ),
            ),
            (
                rep.is_mintable,
                Severity.WARN,
                20,
                "Supply is mintable",
                ("GoPlus reports is_mintable=1: the supply can be inflated, diluting holders."),
            ),
            (
                rep.transfer_pausable,
                Severity.WARN,
                25,
                "Transfers can be paused",
                ("GoPlus reports transfer_pausable=1: an admin can freeze all transfers, trapping your position."),
            ),
            (
                rep.is_blacklisted,
                Severity.WARN,
                20,
                "Blacklist function present",
                ("GoPlus reports is_blacklisted=1: specific addresses can be blocked from trading."),
            ),
        ]
        for present, severity, score, title, detail in flags:
            if present:
                out.append(
                    CheckResult(
                        check=self.id,
                        severity=severity,
                        score=score,
                        title=title,
                        detail=detail,
                        evidence={"source": "goplus"},
                    )
                )
        if not out:
            out.append(
                CheckResult(
                    check=self.id,
                    severity=Severity.OK,
                    score=0,
                    title="No dangerous admin permissions",
                    detail="GoPlus found no mint / pause / blacklist / balance-change powers.",
                    evidence={"source": "goplus"},
                )
            )
        return out


class ReputationSourceCheck:
    """Verified-source and holder-count context from GoPlus."""

    id = "REP-SOURCE"

    async def run(self, ctx: Context) -> list[CheckResult]:
        rep = _report(ctx)
        if rep is None:
            return []
        out: list[CheckResult] = []

        if rep.is_open_source is False:
            out.append(
                CheckResult(
                    check=self.id,
                    severity=Severity.WARN,
                    score=20,
                    title="Contract source is not verified",
                    detail=(
                        "GoPlus reports is_open_source=0: the contract's source is not published, "
                        "so its behavior cannot be reviewed. Treat unverified tokens with caution."
                    ),
                    evidence={"source": "goplus"},
                )
            )
        elif rep.is_open_source is True:
            out.append(
                CheckResult(
                    check=self.id,
                    severity=Severity.OK,
                    score=0,
                    title="Source verified",
                    detail="GoPlus reports the contract source is open/verified.",
                    evidence={"source": "goplus"},
                )
            )

        if rep.holder_count is not None:
            sev = Severity.WARN if rep.holder_count < 50 else Severity.INFO
            score = 15 if rep.holder_count < 50 else 0
            out.append(
                CheckResult(
                    check=self.id,
                    severity=sev,
                    score=score,
                    title=f"{rep.holder_count:,} holders",
                    detail=(
                        f"GoPlus reports {rep.holder_count:,} holders."
                        + (" Very few holders concentrate exit risk." if rep.holder_count < 50 else "")
                    ),
                    evidence={"source": "goplus", "holder_count": str(rep.holder_count)},
                )
            )
        return out

File: packages/scanner/src/checks/reputation.ts
/**
 * Reputation checks backed by the GoPlus token-security dataset. Ported from
 * Hood Trade (MIT). These corroborate the on-chain simulation with a
 * third-party label set: honeypot / cannot-sell-all, dangerous owner
 * permissions, and source-verification / proxy signals.
 */

import type { CheckResult, TokenSnapshot } from "../types.js";

export function reputationHoneypot(s: TokenSnapshot): CheckResult[] {
  const g = s.goPlus;
  if (!g?.found) return [];
  const out: CheckResult[] = [];
  if (g.isHoneypot || g.cannotSellAll) {
    out.push({
      check: "REP-HONEYPOT",
      severity: "danger",
      score: 85,
      title: "Flagged as a honeypot by GoPlus",
      detail: "The GoPlus security dataset labels this token as a honeypot or 'cannot sell all'. Independent corroboration of the on-chain simulation.",
      evidence: { is_honeypot: String(!!g.isHoneypot), cannot_sell_all: String(!!g.cannotSellAll) },
    });
  }
  const sellTax = g.sellTax ?? 0;
  const buyTax = g.buyTax ?? 0;
  if (sellTax >= 0.5 || buyTax >= 0.5) {
    out.push({
      check: "REP-TAX",
      severity: "danger",
      score: 40,
      title: `Punitive transfer tax (${pct(Math.max(sellTax, buyTax))})`,
      detail: "Buy or sell tax is high enough to erase most of the position on a round trip.",
      evidence: { buy_tax: pct(buyTax), sell_tax: pct(sellTax) },
    });
  } else if (sellTax >= 0.1 || buyTax >= 0.1) {
    out.push({
      check: "REP-TAX",
      severity: "warn",
      score: 15,
      title: `Notable transfer tax (${pct(Math.max(sellTax, buyTax))})`,
      detail: "There is a meaningful buy/sell tax. Factor it into expected slippage.",
      evidence: { buy_tax: pct(buyTax), sell_tax: pct(sellTax) },
    });
  }
  if (!out.length) out.push({ check: "REP-HONEYPOT", severity: "ok", score: 0, title: "No honeypot label", detail: "GoPlus does not flag this token as a honeypot." });
  return out;
}

export function reputationPermissions(s: TokenSnapshot): CheckResult[] {
  const g = s.goPlus;
  if (!g?.found) return [];
  const out: CheckResult[] = [];
  if (g.ownerChangeBalance) out.push(danger("REP-BALANCE", 60, "Owner can change balances", "The contract lets a privileged address rewrite holder balances — a direct rug primitive."));
  if (g.isMintable) out.push(warn("REP-MINT", 20, "Token is mintable", "New supply can be minted, diluting holders. Confirm mint authority is renounced or time-locked."));
  if (g.transferPausable) out.push(warn("REP-PAUSE", 18, "Transfers are pausable", "A privileged address can pause all transfers, freezing your position."));
  if (g.hasBlacklist) out.push(warn("REP-BLACKLIST", 18, "Blacklist function present", "The contract can block specific addresses from transferring."));
  if (!out.length) out.push({ check: "REP-PERMISSIONS", severity: "ok", score: 0, title: "No dangerous owner permissions", detail: "No mint / pause / balance-rewrite / blacklist powers detected." });
  return out;
}

export function reputationSource(s: TokenSnapshot): CheckResult[] {
  const g = s.goPlus;
  if (!g?.found) return [];
  const out: CheckResult[] = [];
  if (g.isOpenSource === false) out.push(warn("REP-SOURCE", 15, "Source code not verified", "The contract source is not verified, so its behavior can't be audited from bytecode alone."));
  if (g.isProxy) out.push(warn("REP-PROXY", 12, "Upgradeable proxy", "The token is a proxy; its logic can be swapped by whoever controls the admin key."));
  if (!out.length) out.push({ check: "REP-SOURCE", severity: "ok", score: 0, title: "Verified, non-proxy contract", detail: "Source is verified and the contract is not an upgradeable proxy." });
  return out;
}

function danger(check: string, score: number, title: string, detail: string): CheckResult {
  return { check, severity: "danger", score, title, detail };
}
function warn(check: string, score: number, title: string, detail: string): CheckResult {
  return { check, severity: "warn", score, title, detail };
}
function pct(r: number): string {
  return `${Math.round(r * 100)}%`;
}
